Feat/ga messenger payload

mad-nuts opened the pull request:
Refs iris-connect/iris-backlog#258

mad-nuts committed:
feat: add first draft of iris-message list (mockup only)

mad-nuts committed:
feat: improve first draft of iris-message list (mockup only)

mad-nuts committed:
feat: add first draft for message details view (mockup only)

mad-nuts committed:
feat: add first draft for message creation view (mockup only)

mad-nuts committed:
feat: add iris message controller, db-tables, etc. - work in progress

mad-nuts committed:
feat: add iris-message creation (work in progress) and list view with indexed search

mad-nuts committed:
feat: add iris-message-details view and edit functionality (set isRead) and improve message creation by providing dummy contacts, fix: data-table pagination

mad-nuts committed:
feat: enhance routing when navigating back to message list from message details

mad-nuts committed:
feat: add attachments upload & download to iris messages

mad-nuts committed:
feat: fetch message recipients by EPS client

mad-nuts committed:
feat: add functionality for inbound messages

mad-nuts committed:
feat: add tests for iris-messages

mad-nuts committed:
feat: add e2e tests and data normalizer for iris-messages

mad-nuts committed:
fix: add message dummy bridge to avoid circular dependency caused by dummy functionality

mad-nuts committed:
feat: add bff controller tests for iris messages

mad-nuts committed:
feat: add SQL dialects to iris messages + improve folder initialization

mad-nuts committed:
feat: add database tests for iris-messages

mad-nuts committed:
feat: add eps controller tests for iris-messages

mad-nuts committed:
feat: add service tests for iris-messages (wip)

mad-nuts committed:
feat: iris-message: allow empty files (e.g. text files without content)

mad-nuts committed:
feat: add service tests for iris-messages

mad-nuts committed:
feat: add message builder tests for iris-messages

mad-nuts committed:
fix: fix message file column type / length

mad-nuts committed:
feat: enhance content-type handling and add file type restrictions for iris-message attachments

mad-nuts committed:
fix: fix iris-message test data

mad-nuts committed:
feat: add zipCode search for iris-message-contacts / health departments

mad-nuts committed:
fix: apply changes in the rpc client configuration to the EPSIrisMessageClient

mad-nuts committed:
fix: fix e2e tests for iris-messages

mad-nuts committed:
feat: disable file attachments functionality (but leave it in code for later reactivation)

mad-nuts committed:
feat: rename Attachment to FileAttachment to be able to distinguish different attachment types

mad-nuts committed:
feat: add data attachments to iris messages (draft / work in progress)

mad-nuts committed:
feat: add data selection & preview to iris messages (work in progress)

mad-nuts committed:
feat: add validation to iris message data and restrict the data processing to payload only (work in progress)

mad-nuts committed:
feat: defuse iris message data

mad-nuts committed:
feat: add selection element for iris message data + refactoring

mad-nuts committed:
chore: refactoring

mad-nuts committed:
refactor: replace vuex based api with generic observable (work in progress)

mad-nuts committed:
refactor: rework event-tracking-list.view using data-query-handler

mad-nuts committed:
refactor: use data-query-handler in iris-message-list.view

mad-nuts committed:
refactor: iris-messages: replace vuex based api with generic observable

mad-nuts committed:
fix: fix some minor migration issues with new api functionality

mad-nuts committed:
fix: message create store

mad-nuts committed:
feat: add message data import -> update functionality (work in progress)

mad-nuts committed:
feat: add message data import update selection (work in progress)

mad-nuts committed:
chore: some refactoring

mad-nuts committed:
Merge branch 'develop' into feat/ga-messenger-payload

# Conflicts:
#	iris-client-bff/src/main/java/iris/client_bff/config/DataSubmissionConfig.java
#	iris-client-bff/src/main/java/iris/client_bff/config/HibernateSearchConfig.java
#	iris-client-bff/src/main/java/iris/client_bff/events/web/EventDataRequestController.java
#	iris-client-bff/src/main/java/iris/client_bff/iris_messages/IrisMessage.java
#	iris-client-bff/src/main/java/iris/client_bff/iris_messages/IrisMessageFolder.java
#	iris-client-bff/src/main/java/iris/client_bff/iris_messages/IrisMessageFolderInitializer.java
#	iris-client-bff/src/main/java/iris/client_bff/iris_messages/IrisMessageFolderRepository.java
#	iris-client-bff/src/main/java/iris/client_bff/iris_messages/IrisMessageHdContact.java
#	iris-client-bff/src/main/java/iris/client_bff/iris_messages/IrisMessageRepository.java
#	iris-client-bff/src/main/java/iris/client_bff/iris_messages/IrisMessageService.java
#	iris-client-bff/src/main/java/iris/client_bff/iris_messages/eps/EPSIrisMessageClient.java
#	iris-client-bff/src/main/java/iris/client_bff/iris_messages/eps/IrisMessageDataController.java
#	iris-client-bff/src/main/java/iris/client_bff/iris_messages/eps/IrisMessageDataControllerImpl.java
#	iris-client-bff/src/main/java/iris/client_bff/iris_messages/eps/IrisMessageTransferDefuse.java
#	iris-client-bff/src/main/java/iris/client_bff/iris_messages/web/IrisMessageController.java
#	iris-client-bff/src/main/java/iris/client_bff/iris_messages/web/IrisMessageDetailsDto.java
#	iris-client-bff/src/main/java/iris/client_bff/iris_messages/web/IrisMessageFolderDto.java
#	iris-client-bff/src/main/java/iris/client_bff/iris_messages/web/IrisMessageListItemDto.java
#	iris-client-bff/src/test/java/iris/client_bff/dbms/DatabasesystemIT.java
#	iris-client-bff/src/test/java/iris/client_bff/iris_messages/IrisMessageDataInitializer.java
#	iris-client-bff/src/test/java/iris/client_bff/iris_messages/IrisMessageServiceTest.java
#	iris-client-bff/src/test/java/iris/client_bff/iris_messages/IrisMessageTestData.java
#	iris-client-bff/src/test/java/iris/client_bff/iris_messages/eps/IrisMessageDataControllerTest.java
#	iris-client-bff/src/test/java/iris/client_bff/iris_messages/web/IrisMessageControllerTest.java
#	iris-client-fe/src/api/api.ts
#	iris-client-fe/src/api/common.ts
#	iris-client-fe/src/components/data-tree/data-tree.vue
#	iris-client-fe/src/components/sortable-data-table.vue
#	iris-client-fe/src/server/data/dummy-iris-messages.ts
#	iris-client-fe/src/server/mockAPIServer.ts
#	iris-client-fe/src/store/store.config.ts
#	iris-client-fe/src/store/types.ts
#	iris-client-fe/src/utils/pagination.ts
#	iris-client-fe/src/views/iris-message-create/iris-message-create.store.ts
#	iris-client-fe/src/views/iris-message-create/iris-message-create.view.vue
#	iris-client-fe/src/views/iris-message-details/iris-message-details.data.ts
#	iris-client-fe/src/views/iris-message-details/iris-message-details.view.vue
#	iris-client-fe/src/views/iris-message-list/components/iris-message-data-table.vue
#	iris-client-fe/src/views/iris-message-list/components/iris-message-folders-data-tree.vue
#	iris-client-fe/src/views/iris-message-list/components/iris-message-list-nav-link.vue
#	iris-client-fe/src/views/iris-message-list/iris-message-list.data.ts
#	iris-client-fe/src/views/iris-message-list/iris-message-list.view.vue
#	iris-client-fe/tests/e2e/specs/iris-messages.js
#	iris-client-fe/tests/e2e/support/commands.js

mad-nuts committed:
chore: bring message-payload feature branch up to date with develop

mad-nuts committed:
chore: bring message-payload feature branch up to date with develop (wip)

mad-nuts committed:
fix: refactor message data code structure and fix message payload deserialization

mad-nuts committed:
fix: handle missing location in EventDataRequestController

mad-nuts committed:
feat: add duplicates check for message data import + refactoring (wip)

mad-nuts committed:
refactor: enhance form validation and use similar code structure for message data import and export

mad-nuts committed:
Merge branch 'develop' into feat/ga-messenger-payload

# Conflicts:
#	iris-client-fe/src/api/api.ts
#	iris-client-fe/src/api/common.ts
#	iris-client-fe/src/components/pageable/data-query-handler.vue
#	iris-client-fe/src/components/sortable-data-table.vue
#	iris-client-fe/src/server/mockAPIServer.ts
#	iris-client-fe/src/utils/data-export/common.ts
#	iris-client-fe/src/utils/data.ts
#	iris-client-fe/src/utils/pagination.ts
#	iris-client-fe/src/views/event-tracking-details/event-tracking-details.view.vue
#	iris-client-fe/src/views/event-tracking-list/event-tracking-list.view.vue
#	iris-client-fe/src/views/iris-message-details/iris-message-details.data.ts
#	iris-client-fe/src/views/iris-message-list/components/iris-message-list-nav-link.vue
#	iris-client-fe/src/views/iris-message-list/iris-message-list.store.ts
#	iris-client-fe/src/views/iris-message-list/iris-message-list.view.vue

mad-nuts committed:
chore: bring message-payload feature branch up to date with develop

mad-nuts committed:
feat: add message-data database migration files & e2e test store for accessing application store / data

mad-nuts committed:
fix: fix app-bar & iris-message unread message count e2e tests

mad-nuts committed:
refactor: improve e2e tests and data type

mad-nuts committed:
feat: add mockup api for message data attachments

mad-nuts committed:
Merge branch 'develop' into feat/ga-messenger-payload

mad-nuts committed:
feat: update db iris-message-data migration files, add message data attachment count (total & imported) and add e2e tests for iris-message-data

mad-nuts committed:
fix: fix iris-message badge & e2e tests

mad-nuts committed:
chore: cleanup iris-message-data e2e tests

mad-nuts committed:
refactor: move DataAttachmentCountDto to separate file

jekutzsche commented on iris-client-bff/src/main/java/iris/client_bff/iris_messages/web/IrisMessageController.java:
```diff
@@ -181,4 +208,11 @@ private void validateField(String value, String field) {
+
+	private void validateMessageDataPayload(String value, String field) {
+		if (validationHelper.isPossibleAttackForMessageDataPayload(value, field, false)) {
+			throw new ResponseStatusException(HttpStatus.BAD_REQUEST, messages.getMessage("iris_message.invalid_message_data") + ": "+ field);
```
You can use `messages.getMessage("iris_message.invalid_message_data", field)` with `iris_message.invalid_message_data=…: {0}` to avoid string concatenation.

mad-nuts replied:
removed the field part as it is not really necessary

jekutzsche reviewed (changes requested):
The DB scripts are not yet correct, because the DB tests fail during the build.

I'm not finished yet, but would like to give some feedback in between.

mad-nuts commented on iris-client-bff/src/main/java/iris/client_bff/iris_messages/IrisMessageDataProcessors.java:
```diff
@@ -0,0 +1,57 @@
+@Service
+public class IrisMessageDataProcessors {
+
+	private IrisMessageDataRepository irisMessageDataRepository;
+
+	@Autowired
+	public void setIrisMessageDataRepository(IrisMessageDataRepository irisMessageDataRepository) {
+		this.irisMessageDataRepository = irisMessageDataRepository;
+	}
+
+	private MessageSourceAccessor messages;
+
+	@Autowired
+	public void setMessageSourceAccessor(MessageSourceAccessor messages) {
+		this.messages = messages;
+	}
```
The IrisMessageDataProcessors` constructor is autowired itself (line 33), "auto injecting" the IrisMessageDataProcessor list.

The IrisMessageDataProcessor is the interface for all MessageDataProcessor "types", e.g. the EventMessageDataProcessor.

mad-nuts replied:
I think, that passing all dependencies into the constructor would make the idea behind the code harder to understand. But sure, I'll change it.

jekutzsche replied:
Also now you have to look at the code to understand what is done there. It would be generally useful for understanding to give it a name and to outsource this code to a function (e.g. `extractDiscriminatorProcessorMap`).

mad-nuts commented on iris-client-bff/src/main/java/iris/client_bff/iris_messages/IrisMessageDataProcessors.java:
```diff
@@ -0,0 +1,57 @@
+	public IrisMessageDataProcessor getProcessor(String discriminator) {
+		try {
+			return this.processors.get(discriminator);
+		} catch (Throwable e) {
```
Shouldn't be an issue in this case as the error (message) is not used / processed.
But good to know, thanks! I'll change the Throwables to Exceptions.

jekutzsche replied:
Here in the example there will be no problem, but for clarification a few words. The problem is not the usage of the error or message, but the control flow. If there is an `OutOfMemoryError` in the `try{...}`, then it will be caught with `Throwable` too. But there is no meaningful handling without memory (throwing another exception), the application is dead.

mad-nuts committed:
Merge branch 'develop' into feat/ga-messenger-payload

# Conflicts:
#	iris-client-bff/src/main/java/iris/client_bff/iris_messages/eps/EPSIrisMessageClient.java

mad-nuts commented on iris-client-bff/src/main/java/iris/client_bff/events/message/EventMessageDataPayload.java:
```diff
@@ -0,0 +1,95 @@
+				List<String> guestIds) {
+			ModelMapper mapper = new ModelMapper();
+			List<Guest> guests = eventDataSubmission.getGuests().stream()
+					.filter((guest -> guestIds.size() <= 0 || guestIds.contains(guest.getGuestId().toString())))
```
will be changed, thanks

mad-nuts commented on iris-client-bff/src/main/java/iris/client_bff/events/message/EventMessageDataPayload.java:
```diff
@@ -0,0 +1,95 @@
+						mapped.setGuestId(null);
+						mapped.setMessageDataSelectId(UUID.randomUUID().toString());
+						return mapped;
+					}).collect(Collectors.toList());
```
will be changed, thanks

mad-nuts commented on iris-client-bff/src/main/java/iris/client_bff/iris_messages/web/IrisMessageInsertDto.java:
```diff
@@ -22,4 +28,23 @@
     @Size(max = IrisMessage.BODY_MAX_LENGTH)
     private String body;
 
+    @Valid
+    private List<DataAttachment> dataAttachments;
```
will be changed, thanks

mad-nuts commented on iris-client-bff/src/main/resources/db/migration/V1012__add_iris_message_data.sql:
```diff
@@ -0,0 +1,15 @@
+    id uuid NOT NULL,
+    message_id uuid NOT NULL,
+    discriminator varchar(255) NOT NULL,
+    payload varchar(999999) NOT NULL,
```
will be changed to text

mad-nuts commented on iris-client-bff/src/main/java/iris/client_bff/events/message/EventMessageDataPayload.java:
```diff
@@ -0,0 +1,95 @@
+	@Data
+	static class EventDataRequestPayload {
+
+		private String refId;
```
The refId is a required field in the EventDataRequest builder. I set the import value to "-" and exclude the refId from the payload.

jekutzsche replied:
`-` at the beginning is one of the forbidden characters. Not that we get then a 400, e.g. with an update. Maybe rather a text like "nicht vergeben - wurde importiert" with I18N?

mad-nuts replied:
Good idea! Thanks!

mad-nuts commented on iris-client-bff/src/main/java/iris/client_bff/iris_messages/utils/RawJsonDeserializer.java:
```diff
@@ -0,0 +1,18 @@
+
+import java.io.IOException;
+
+public class RawJsonDeserializer extends JsonDeserializer<String> {
+	@Override
+	public String deserialize(JsonParser jsonParser, DeserializationContext deserializationContext) throws IOException {
+		ObjectMapper mapper = (ObjectMapper) jsonParser.getCodec();
+		JsonNode node = mapper.readTree(jsonParser);
+		return mapper.writeValueAsString(node);
+	}
+}
```
ok

mad-nuts committed:
chore: handle code review change requests

mad-nuts commented on iris-client-bff/src/main/java/iris/client_bff/events/message/dto/ExportSelectionDto.java:
```diff
@@ -0,0 +1,30 @@
+
+	public static ExportSelectionDto toModel(String payload) {
+		try {
+			ObjectMapper objectMapper = new ObjectMapper();
```
I'll change it to the default ObjectMapper

mad-nuts commented on iris-client-bff/src/main/java/iris/client_bff/events/message/dto/ImportSelectionDto.java:
```diff
@@ -0,0 +1,27 @@
+
+	public static ImportSelectionDto toModel(String payload) {
+		try {
+			ObjectMapper objectMapper = new ObjectMapper();
```
I'll change it to the default ObjectMapper

mad-nuts commented on iris-client-bff/src/main/java/iris/client_bff/events/message/EventMessageDataPayloadDefuse.java:
```diff
@@ -0,0 +1,48 @@
+package iris.client_bff.events.message;
+
+import iris.client_bff.core.utils.ValidationHelper;
+import iris.client_bff.events.eps.EventDataDefuse;
```
I'll move it

mad-nuts committed:
chore: handle code review change requests

mad-nuts committed:
fix: set iris message data payload column type for mysql to "mediumblob" by setting the column length

mad-nuts committed:
Merge branch 'develop' into feat/ga-messenger-payload

# Conflicts:
#	iris-client-bff/src/main/java/iris/client_bff/iris_messages/eps/EPSIrisMessageClient.java

mad-nuts committed:
feat: add some iris message data bff tests

mad-nuts committed:
feat: add some iris message data bff tests

mad-nuts committed:
fix: apply multiple changes based on code review feedback

mad-nuts committed:
fix: apply multiple changes based on code review feedback

mad-nuts committed:
fix: revert changes to object mapper to fix tests

jekutzsche committed:
chore: removes unnecessary and perhaps dangerous configuration changes to the mapper and moves this to the test

Changes sometime at runtime the configuration of the central `ObjectMapper`s and this holds a danger.

Where the new `ObjectMapper` is created in the test, it should also be configured.

jekutzsche committed:
chore: code cleanup

jekutzsche commented:
:tada: This PR is included in version 1.6.0 :tada:

The release is available on [GitHub release](https://github.com/iris-connect/iris-client/releases/tag/v1.6.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
